dss-collection-nft/contracts/DSSCollection.cdc

```diff
@@ -0,0 +1,695 @@
+                self.open == true: "not open"
+            }
+
+            self.open = false
+
+            emit CollectionGroupClosed(id: self.id)
+        }
+
+        // Mint a DSSCollection NFT in this group
+        //
+        pub fun mint(completedBy: String, level: UInt8): @DSSCollection.NFT {
+            pre {
+                self.open != true: "cannot mint an open collection group"
+                DSSCollection.validateTimeRange(
+                    timeBound: self.timeBound,
+                    startTime: self.startTime,
+                    endTime: self.endTime
+                ) == true : "cannot mint a collection group outside of time bounds"
+                level <= 10: "token level must be less than 10"
+            }
+
+            // Create the DSSCollection NFT, filled out with our information
+            //
+            let dssCollectionNFT <- create NFT(
+                collectionGroupID: self.id,
+                serialNumber: self.numMinted + 1,
+                completedBy: completedBy,
+                level: level
+            )
+            DSSCollection.totalSupply = DSSCollection.totalSupply + 1
+            self.numMinted = self.numMinted + 1 as UInt64
+
+            return <- dssCollectionNFT
+        }
+
+        init (
+            name: String,
+            description: String,
+            typeName: String,
+            startTime: UFix64?,
+            endTime: UFix64?,
+            timeBound: Bool
+        ) {
+            self.id = self.uuid
+            self.name = name
+            self.description = description
+            self.typeName = typeName
+            self.open = true
+            self.startTime = startTime
+            self.endTime = endTime
+            self.timeBound = timeBound
+            self.numMinted = 0 as UInt64
+
+            emit CollectionGroupCreated(
+                id: self.id,
+                name: self.name,
+                description: self.description,
+                typeName: self.typeName,
+                startTime: self.startTime,
+                endTime: self.endTime,
+                timeBound: self.timeBound
+            )
+        }
+    }
+
+    // Get the publicly available data for a CollectionGroup by id
+    //
+    pub fun getCollectionGroupData(id: UInt64): DSSCollection.CollectionGroupData {
+        pre {
+            DSSCollection.collectionGroupByID[id] != nil: "Cannot borrow collection group, no such id"
+        }
+
+        return DSSCollection.CollectionGroupData(id: id)
+    }
+
+    // Get the publicly available data for a Slot by id
+    //
+    pub fun getSlotData(id: UInt64): DSSCollection.SlotData {
+        pre {
+            DSSCollection.slotByID[id] != nil: "Cannot borrow slot, no such id"
+        }
+
+        return DSSCollection.SlotData(id: id)
+    }
+
+    // Get the publicly available data for a Slot by id
+    //
+    pub fun getItem(id: UInt64): DSSCollection.Item {
+        return DSSCollection.itemByID[id]!
```

Comment:
It can panic if the given `Id` doesn't exists, It is better to give a failure message.

```
return DSSCollection.itemByID[id] ?? panic("Given Id Item doesn't exists")
```

Reply:
Thanks Satya. I ended up removing this function.